sim/lib/dynamics.py

```diff
@@ -666,19 +670,32 @@ def __push_contact_exposure_events(self, t, infector, base_rate):
         of person `i` (equivalent to `\mu` in model definition)
         """
 
-        def valid_j():
-            '''Generates indices j where `infector` is present
-            at least `self.delta` hours before j '''
-            for j in range(self.n_people):
-                if self.state['susc'][j]:
-                    if self.mob.will_be_in_contact(indiv_i=j, indiv_j=infector, t=t, site=None):
-                        yield j
+        if not self.dynamic_tracing:
+            def valid_j():
+                '''Generates indices j where `infector` is present
+                at least `self.delta` hours before j '''
+                for j in range(self.n_people):
+                    if self.state['susc'][j]:
+                        if self.mob.will_be_in_contact(indiv_i=j, indiv_j=infector, t=t, site=None):
+                            yield j
+
+            valid_contacts = valid_j()
+        else:
+            infectors_contacts = self.mob.find_contacts_of_indiv(self.mob.all_mob_traces, indiv=infector, tmin=t)
+
+            valid_contacts = []
+            for contact in infectors_contacts:
+                if self.state['susc'][contact.indiv_i]:
+                    if contact not in self.mob.contacts[contact.indiv_i][infector]:
+                        self.mob.contacts[contact.indiv_i][infector].update([contact])
```

Comment:
Is it correct that we check/update `[contact.indiv_i][infector]` and not `[infector][contact.indiv_i]`? Need to make sure is consistent with `delta` contact overlap direction

Reply:
Yes I think it is quite confusing in the code because mob.contact[i][j] is defined (w.r.t. the dela contact) as the contact of a possibly infectious person j to a person i, I agree it would be more intuitive to define the first index as corresponding to the possible infector but I didn't want to make to many changes in the existing code. 

Later in the valid_j() function the person j is called infector and the possibly susceptible person i is called j:
self.mob.will_be_in_contact(indiv_i=j, indiv_j=infector, t=t, site=None)

This actually caused a few bugs.
I would suggest to either reverse the order in the contact array or to swap the naming of the variables later on to avoid things like (indiv_i=j, indiv_j=i).

There would be quite a lot of places where it would have to be changed but it would make it much more comprehensible. If you agree I could make the changes and send another pull request.

Reply:
It may be safer to just rename the variables rather than reverse the order. But I agree that the current naming may lead to confusion. I let @larslorch to decide since he is taking the lead on this code review.

Reply:
I think we should leave the calls to `self.mob.will_be_in_contact(indiv_i=j, indiv_j=infector, t=t, site=None)` as they are currently, I think we are all used to the direction/order of the call now and they are correct the way they are currently done.
With my comment I just wanted to make sure to double check that the newly added line does the correct thing. Thanks!
@HeinerKremer @manuelgr